router/cmd/main.go

```diff
@@ -71,59 +106,148 @@ func Main() {
+				zap.String("config_file", config.DefaultConfigPath),
+			)
+		} else {
+			logger.Info(
+				"Config file path provided. Values in the config file have higher priority than environment variables",
+				zap.String("config_file", configPath),
+			)
+		}
 	}
 
-	// Provide a way to cancel all running components of the router after graceful shutdown
-	// Don't use the parent context that is canceled by the signal handler
-	routerCtx, routerCancel := context.WithCancel(context.Background())
-	defer routerCancel()
-
-	router, err := NewRouter(routerCtx, Params{
-		Config: &result.Config,
-		Logger: logger,
-	})
-	if err != nil {
-		logger.Fatal("Could not create router", zap.Error(err))
-	}
+	reloadChan := make(chan os.Signal, 1)
 
-	if err = router.Start(routerCtx); err != nil {
-		logger.Fatal("Could not start router", zap.Error(err))
-	}
+	signal.Notify(reloadChan, syscall.SIGHUP)
 
-	<-ctx.Done()
+	// Setup config file watcher if enabled
+	if result.Config.WatchConfig.Enabled {
+		startupDelay := 0 * time.Second
 
-	logger.Info("Graceful shutdown of router initiated", zap.String("shutdown_delay", result.Config.ShutdownDelay.String()))
+		// Apply startup delay if configured
+		if result.Config.WatchConfig.StartupDelay.Enabled {
+			startupDelay = timex.RandomDuration(result.Config.WatchConfig.StartupDelay.Maximum)
 
-	// Enforce a maximum shutdown delay to avoid waiting forever
-	// Don't use the parent context that is canceled by the signal handler
-	shutdownCtx, cancel := context.WithTimeout(context.Background(), result.Config.ShutdownDelay)
-	defer cancel()
+			logger.Info("Using startup delay before initializing config watcher",
+				zap.Duration("delay", startupDelay),
+			)
+		}
 
-	if err = router.Shutdown(shutdownCtx); err != nil {
-		if errors.Is(err, context.DeadlineExceeded) {
-			logger.Warn("Router shutdown deadline exceeded. Consider increasing the shutdown delay")
+		watchFunc, err := watcher.New(watcher.Options{
+			Interval: result.Config.WatchConfig.Interval,
+			Logger:   logger.With(zap.String("watcher_label", "router_config")),
+			Path:     configPath,
+			Callback: func() {
+				logger.Debug("Configuration changed, triggering reload")
+
+				// Just a hack to make channel code simpler
+				reloadChan <- syscall.SIGHUP
+			},
+		})
+		if err != nil {
+			logger.Error("Could not create watcher", zap.Error(err))
+			return
 		}
-		logger.Fatal("Could not shutdown router gracefully", zap.Error(err))
+
+		watcherCtx, watcherCancel := context.WithCancel(context.Background())
+		defer watcherCancel()
+
+		go func() {
+			// Sleep for startupDelay to prevent synchronized reloads across
+			// different instances of the router
+			time.Sleep(startupDelay)
+
+			if err := watchFunc(watcherCtx); err != nil {
+				if err != context.Canceled {
+					logger.Error("Error watching execution config", zap.Error(err))
+				}
+			}
+		}()
+
+		logger.Info("Watching router config file",
```

Comment:
This needs to be split into multiple files. The reloader should be split from the actual router start logic.

Reply:
What do you mean? You'd like a general purpose command restarter? It needs to be in the same place as the router start logic in order to control it.

Reply:
Both the file watcher and SIGHUP just send a signal to a new channel `reloadChan`, which sits next to the old `ctx.Done()` as possible ways to trigger a router shutdown, with the former not setting a flag that causes process exit at the end of the graceful shutdown.

Reply:
It would be great if we could abstract that a little bit better away without bloating up the main file. The file watcher code should not run in production.